sql-plugin/src/main/scala/com/nvidia/spark/rapids/PrioritySemaphore.scala

```diff
@@ -58,10 +58,21 @@ class PrioritySemaphore[T](val maxPermits: Int)(implicit ordering: Ordering[T])
     lock.lock()
     try {
       val condition = lock.newCondition()
+      var queued = false
       while (!canAcquire(numPermits)) {
```

Comment:
There's a question here of what to do in this scenario:
- There's 10 permits in the semaphore, 9 are taken.
- The front of the queue wants 2 permits, which is why it is blocked
- Another thread comes in with only 1 permit but is less desired by priority than the front of the queue

Currently the code lets this thread through, effectively creating a priority inversion.  Intentional?

Reply:
I considered that but didn't really worry about it, mainly because IIUC we currently don't actually use dynamic sizing (so everything has the same number of permits).

Reply:
It's easy to solve the priority inversion by comparing the current priority against the priority at the head of the queue.  That brings up another issue, which is tryAcquire and acquire don't have the same signature.  The only difference between them should be whether it blocks.

Reply:
Agreed, but it also brings up another question which is, might we actually want to let lower prio tasks run if they have fewer permits such that they can "fit" on the gpu whereas the head of the line currently cannot fit? This would probably be beneficial in certain cases, but it would also be more complex.

Reply:
I just realized I'm basically just saying the same thing as you said in the top of the thread.... so yes the question comes down to what is the intent.. to me it makes sense to basically "solve the priority inversion" as you laid out and for now assume we never want to invert priority in order to run smaller tasks. We can revisit that later if needed.

Reply:
> might we actually want to let lower prio tasks run if they have fewer permits such that they can "fit"

I would argue the permit request size should factor into the calculation of the priority value if that's really the behavior we want, otherwise priority isn't really a priority.

Reply:
Sure that is reasonable.